### PR TITLE
Allow users to configure compression levels

### PR DESCRIPTION
**Pull request by dralley**
### 📜 Checklist

- [x] Commits are cleanly separated and have useful messages
- [x] A changelog entry or entries has been added to CHANGELOG.md
- [x] Documentation is thorough
- [x] Test coverage is excellent and passes
- [x] Works when tests are run `--all-features` enabled

### CODE REVIEW

**Comment by dralley** [comment]
@ikrivosheev

**Comment by dralley** [comment]
About the default compression type: I'm open to changing it, but I really do think we need to opt into some kind of compression by default.  I'm in a position to see some of the infrastructure hosting costs of large repos, and have spoken with users that manage terabytes of RPMs, so to me the cost/benefit ratio of the various defaults is very stark in favor of some kind of compression.

Zstd would be ideal but so long as CentOS 7 is our minimum bound we cannot choose it.  So until then, Gzip is probably the best tradeoff.

File: src/rpm/builder.rs [hunk 1]
```diff
@@ -168,7 +168,7 @@ impl RPMBuilder {
     ///
     /// ```
     /// # fn foo() -> Result<(), Box<dyn std::error::Error>> {
-    /// let pkg = rpm::RPMBuilder::new("foo", "1.0.0", "MPL", "x86_64", "some bar package")
+    /// let pkg = rpm::RPMBuilder::new("foo", "1.0.0", "MIT", "x86_64", "some bar package")
```
Comment (drahnr) [inline src/rpm/builder.rs:171]: That wasn't a typo :)

File: src/rpm/builder.rs [hunk 2]
```diff
@@ -198,28 +198,62 @@ impl RPMBuilder {
...
+    ///
+    /// The a changelog entry consists of an entry name (which includes author, email followed by
+    /// a dash followed by a version number), description, and the date and time of the change.
 
     /// ```
     /// # fn foo() -> Result<(), Box<dyn std::error::Error>> {
     /// use rpm::chrono::TimeZone;
     ///
-    /// let pkg = rpm::RPMBuilder::new("foo", "1.0.0", "Apache-2.0", "x86_64", "some baz package")
+    /// let pkg = rpm::RPMBuilder::new("foo", "1.0.0", "MIT", "x86_64", "some baz package")
```
Comment (drahnr) [inline src/rpm/builder.rs:256]: I don't understand why all examples are unified to MIT?

**Review by drahnr — CHANGES_REQUESTED** [review]
Besides the two odd stray changes, LGTM

File: src/rpm/builder.rs (hunk 2 above)
Comment (dralley) [reply src/rpm/builder.rs:256]: It's just shorter

File: src/rpm/builder.rs (hunk 1 above)
Comment (dralley) [reply src/rpm/builder.rs:171]: It's not a correct SPDX identifier though, I could change it to `MPL-2.0`

Comment (drahnr) [reply src/rpm/builder.rs:171]: Yes please

File: src/rpm/builder.rs (hunk 2 above)
Comment (drahnr) [reply src/rpm/builder.rs:256]: Please revert

**Comment by dralley** [comment]
Wdyt about gzip vs. zstd as the default?  Knowing that the latter can't be used on EL7

**Comment by drahnr** [comment]
I am still leaning towards `None`, but I don't have a strong opinion. It'd be nice to at least not break anything that's currently officially supported/maintained by default to avoid any surprises.

File: src/rpm/compressor.rs [hunk 3]
```diff
@@ -73,32 +75,45 @@ impl Compressor {
...
-            Compressor::Gzip(encoder) => Ok(encoder.finish().into_result()?),
+            Compressor::Gzip(encoder) => Ok(encoder.finish()?),
             Compressor::Zstd(encoder) => Ok(encoder.finish()?),
             Compressor::Xz(encoder) => Ok(encoder.finish()?),
         }
     }
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum CompressionDetails {
```
Comment (dralley) [inline src/rpm/compressor.rs:86]: Should this be named `CompressionWithLevel`

File: tests/compat.rs [hunk 4]
```diff
@@ -222,7 +222,7 @@ mod pgp {
                 "x86_64",
                 "spins round and round",
             )
-            .compression(CompressionType::Gzip)
+            .compression(CompressionType::Zstd)
```
Comment (dralley) [inline tests/compat.rs:225]: Just trying to increase coverage on the compat tests, we use Gzip already on one of them.

**Review by ikrivosheev — COMMENTED** [review]
@dralley hello! It looks good, but I don't see an API for set compression level in `Compressor`.

**Comment by dralley** [comment]
@ikrivosheev  See the 2nd example provided in the documentation

**Comment by ikrivosheev** [comment]
@dralley for me it looks strange. We have two public types for one operation. Users need to think about it.

**Comment by dralley** [comment]
Imo it's not so different from APIs which take AsRef<str> or Into<Cow<str>>

**Comment by dralley** [comment]
>I think is big different between AsRef and this API. When you use AsRef this mean that you want to work with objects like for example: Vec<u8>, [u8], Mmap, ... These objects are not similar but they have a similar interface.    Here you have a very similar objects.

I don't see that as a big difference.  Besides, a `&str` and `String` and a `Path` and a `PathBuf` are very similar objects and they will all work in an `AsRef<Path>` parameter

**Comment by ikrivosheev** [comment]
> > I think is big different between AsRef and this API. When you use AsRef this mean that you want to work with objects like for example: Vec, [u8], Mmap, ... These objects are not similar but they have a similar interface.    Here you have a very similar objects.
> 
> I don't see that as a big difference. Besides, a `&str` and `String` and a `Path` and a `PathBuf` are very similar objects and they will all work in an `AsRef<Path>` parameter

We need another opinion)

**Review by drahnr — APPROVED** [review]
I agree that this is not the must idiomatic interface, generally the decompressor doesn't care about it much, for the compressing side only the actual compression `fn` _must_ know about it. Now the need to provide the compression level as header causes the whole dependency and I don't think there is _great_ solution for it given that requirement.

The impl is a _good_ solution :+1: 

---

`AsRef<Path>` liberates the caller from doing extra conversions, so I am all for that.

**Pull request merged**